### PR TITLE
fix(celery): Implement Kafka producer shutdown correctly

### PR DESCRIPTION
**Pull request by untitaker**
There are a few uses of atexit in our codebase to reuse an object like a
Kafka producer across multiple tasks. It turns out that atexit works in
only a few circumstances, even in Celery 5. This particularly
matters in two cases:

1. an exception ending the worker process
2. celery main process deciding to reap the subprocess via SIGTERM (such
   as when hitting max_tasks_per_child) -- this actually happens very
   often.

It is more reliable to patch the worker loop to run shutdown reliably,
like we do in the SDK.

I tested this change manually and have a rough idea how to write an
automated test for it, just putting this PR up for discussion for now.

### CODE REVIEW

**Review by Swatinem — APPROVED** [review]

File: src/sentry/usage_accountant/accountant.py [hunk 1]
```diff
@@ -82,6 +84,6 @@ def record(
         )
 
         _accountant_backend = UsageAccumulator(producer=producer)
-        atexit.register(_shutdown)
+        register_shutdown(_shutdown)
```
Comment (fpacifici) [inline src/sentry/usage_accountant/accountant.py:87]: I am not sure this will be enough.
This method is not supposed to be celery specific. We need to be able to shutdown the accountant in other environment as well. 

I'd assume your goal is to replace the `atexit` shutdown with the patch you added here. But that is only going to work for celery.

File: src/sentry/utils/arroyo_producer.py [hunk 2]
```diff
@@ -38,7 +39,8 @@ def produce(self, destination: Union[Topic, Partition], payload: KafkaPayload) -
     def _get(self) -> KafkaProducer:
         if self._producer is None:
             self._producer = self._factory()
-            atexit.register(self._shutdown)
+
+            register_shutdown(self._shutdown)
```
Comment (fpacifici) [inline src/sentry/utils/arroyo_producer.py:43]: Isn't this going to work only for producers used inside Celery workers? 

If we make this class Celery specific I think we should make it more explicit in the docstring and with the name.

File: src/sentry/usage_accountant/accountant.py [hunk 3]
```diff
@@ -51,6 +52,7 @@ def _shutdown() -> None:
         _accountant_backend.flush()
         _accountant_backend.close()
         logger.info("Usage accountant flushed and closed.")
+        reset_backend()
```
Comment (fpacifici) [inline src/sentry/usage_accountant/accountant.py:55]: Why this change ?

File: src/sentry/usage_accountant/accountant.py (hunk 1 above)
Comment (untitaker) [reply src/sentry/usage_accountant/accountant.py:87]: we can continue to register it on atexit then, but it remains to be tested whether that actually works on uwsgi

File: src/sentry/usage_accountant/accountant.py (hunk 3 above)
Comment (untitaker) [reply src/sentry/usage_accountant/accountant.py:55]: if people call `_shutdown` multiple times (and I think that will happen especially if we register for both atexit and celery-specific things), then I think we should not close the same backend twice.
